### PR TITLE
[SPARK-42786][Connect] Typed Select

### PR DESCRIPTION
**Pull request by zhenlineo**
### What changes were proposed in this pull request?
Implement typed select methods in the Dataset.

### Why are the changes needed?
More APIs

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Unit and E2E tests

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/AgnosticEncoder.scala [hunk 1]
```diff
@@ -107,6 +108,51 @@ object AgnosticEncoders {
     override def dataType: DataType = schema
   }
 
+  case class TupleEncoder[T](
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/AgnosticEncoder.scala:111]: @zhenlineo we should be able to use `ProductEncoder` for this.

Comment (zhenlineo) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/AgnosticEncoder.scala]: Done. Any special case we missed?

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 2]
```diff
@@ -1084,6 +1084,68 @@ class Dataset[T] private[sql] (
...
   }
 
+  /**
+   * Internal helper function for building typed selects that return tuples. For simplicity and
+   * code reuse, we do this without the help of the type system and then use helper functions that
+   * cast appropriately for the user facing interface.
+   */
+  private def selectUntyped(columns: TypedColumn[_, _]*): Dataset[_] = {
+    val encoder = ProductEncoder.tuple(columns.map(_.encoder))
+    sparkSession.newDataset(encoder) { builder =>
```
Comment (hvanhovell) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala:1094]: Nit you could make this is a bit DRYer if you combine some of the code with select.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/AgnosticEncoder.scala [hunk 3]
```diff
@@ -107,6 +108,16 @@ object AgnosticEncoders {
     override def dataType: DataType = schema
   }
 
+  object ProductEncoder {
+    def tuple(encoders: Seq[AgnosticEncoder[_]]): AgnosticEncoder[_] = {
+      val fields = encoders.zipWithIndex.map {
+        case (e, id) => EncoderField(s"_${id + 1}", e, e.nullable, Metadata.empty)
+      }
+      val cls = Utils.getContextOrSparkClassLoader.loadClass(s"scala.Tuple${encoders.size}")
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/AgnosticEncoder.scala:116]: Can we try to cache this?

**Review by hvanhovell — APPROVED** [review]
LGTM

**Comment by hvanhovell** [comment]
Merging to master!

**Pull request closed without merging**